### ISSUE TITLE
Changing the key of a component unexpectedly reorders the DOM

### ISSUE THREAD

**Issue by esrch** [lhorie/mithril.js#912]
I needed to force the redraw of a component in a view, so I figured that I would simply change the key associated with it. However, changing the key changes the order of elements in the DOM. [Here](http://codepen.io/anon/pen/wMWeEg) is an example of that behavior.

If you press on the button, it moves the button to the end of the DOM. What is even stranger is that the key is not set on the button, but it is the button that moves.

**Comment by syaiful6** [comment]
hmm, look like you need to write your view like this
```
var Subcomponent = {
  view: function(ctrl) {
    return m('span', 'c');
  }
}

var Component = {
  controller: function() {
    ctrl = this;
    this.k = 1;
    this.inc = function() {
      ctrl.k++;
    }
  },
  view: function(ctrl) {
    return m('div', {}, 'a',
      m('span', {}, 'b',
       m.component(Subcomponent, {key: ctrl.k}), 
        m('button', {onclick: ctrl.inc},'+', 
          m('span', {}, ctrl.k))));
  }
}

m.mount(document.body, Component);
```

if your write like that, it will not moved.

**Comment by esrch** [comment]
@syaiful6 

Thanks a lot for the solution. I'm still surprised at the behavior, but at least, there is a way around it.

**Comment by barneycarroll** [comment]
I'm not sure this should be closed — it's still a bug IMO, and the root cause isn't clear.

**Comment by isiahmeadows** [comment]
@barneycarroll You're right. it is most definitely a bug.

**Comment by vicnicius** [comment]
Was facing a similar issue and took a closer look at it. 

What happens is that when you dynamically update the key attribute of an element, you are cheating the key differ algorithm [here](https://github.com/lhorie/mithril.js/blob/next/mithril.js#L137). 

As the key changes, the cached node action index will not be matched when doing the diff, and the algorithm will then add the element as if it didn't exist before. In the example, the algorithm will look for the index + 1 on the cached node actions. That's why the keyed element goes to the end of the tree (as if it was new).

To me, it looks like a desired behavior. It should definitely be documented, though. 

But there is another issue related to the key diffing algorithm, and this one does look like a bug: https://github.com/lhorie/mithril.js/issues/934

**Comment by isiahmeadows** [comment]
@vicnicius This is a different bug than the one you referenced.

**Comment by vicnicius** [comment]
Sure, @isiahmeadows. I just referenced it because it is somewhat related to the same feature. Also to state that IMO while that one (#934) is clearly a bug, this one (#912), if I'm not mistaken in my analysis, is not, and should be closed.

**Comment by isiahmeadows** [comment]
@vicnicius Compare [this pen](http://codepen.io/isiahmeadows/pen/PZQqgX) to [this pen](http://codepen.io/isiahmeadows/pen/gPvpEB) The expected behavior is that the button stays in the middle. That is the bug: components are getting reordered when the last element isn't a key, but I'm thinking that may be very difficult to fix, since text nodes aren't exactly regular DOM nodes.

I would have to come up with an internal tree separate from the DOM, which will increase the size of this library significantly, as well as increase the memory requirements somewhat, since this tree will literally hold a clone of the DOM, and will never be fully garbage collected as long as the page exists. And that's not trivial, as I'm working on such a thing myself in spare time (it's not open yet, since the tree editing algorithm isn't trivial, it's internally structured extremely differently from anything else out there, and I'm [looking at various CS algorithms](http://www.inf.unibz.it/dis/projects/tree-edit-distance/tree-edit-distance.php) for optimization help).

**Comment by vicnicius** [comment]
@isiahmeadows I understand but I still think that it is a feature misuse that is leading the key diffing algorithm to this unexpected behavior. 

What makes example two work is the propagation of the "key diffing algorithm searching wrongly for an index" error to the other keyed elements. The tree is rebuilt on the right way, but by luck: instead of setting the elements right position, it's considering that each one of them is actually a new on, in order, what ends up building a visually correct tree. Add a text node anywhere between the first and the last keyed element and it will break.

On the other way, if we simple keep the key state on the first example, it will work as expected. See example [here](http://codepen.io/vicnicius/pen/qbxNvQ). 

I see this as expected because actually the key differing algorithm is doing what it is supposed to do. It's task is to maintain a set of elements on the same order from one redraw to the next. For this to happen, the components must maintain their identities (keys) to the algorithm from one diffing to the other, so that it knows which is which. Change the key and it will see it as a new element, appending it to the end. Sure it could take care of such cases, but like you said, this would add a lot of unnecessary code and affect performance badly. Actually, if a diffing algorithm is able to take care of such cases, keys would not be needed. 

Now those tree editing algorithms you mentioned do look really amazing! Will take a look at them, and I will love to take a look at your implementation of them when you get to publish it. I've been digging into [incremental-dom](https://github.com/google/incremental-dom) lately and have been very interested in algorithm solutions to tree problems lately.

**Comment by isiahmeadows** [comment]
@vicnicius Back to the original bug, the reason I call it a bug is because this is normal code. Consider [this pen](http://codepen.io/isiahmeadows/pen/VeQPdq) and [this pen](http://codepen.io/isiahmeadows/pen/gPvpEB). Either way, those two shouldn't be flipping. It's unexpected behavior, and [React doesn't have that problem](http://codepen.io/isiahmeadows/pen/GoQrXV).

**Comment by isiahmeadows** [comment]
Also:

1. I can also tell you that incremental-dom was created because Google Closure needed unbalanced templates, and nothing else out there had that ability.

2. You'll have a tremendous perf drop if you don't have keys. The reason is that you have to do an O(mn) search for each element otherwise. Mine will be a little easier to accomplish because I'll have regular components and base components optionally with roots (that each have their own connected tree). There's a reason why React *doesn't* actually use an established tree diffing algorithm verbatim: even though existing tree editing algorithms are nearly optimal for memory, speed is a joke if you're dealing with predictable trees with parts ignored like what's typical with virtual DOMs. Also, if you have a router and you're working on some demo, you probably don't want to keep the same counter you had from another route for something completely different. So, there's logical concerns as well. :smile:

**Comment by leeoniya** [comment]
Not sure how you guys do the diff internally, but in domvm [1] I run through the vtree and `childNodes` in parallel and ensure they're all aligned/hydrated correctly using `insertBefore` rather than `appendChild`. This ensures all reused DOM nodes match the vtree. Maybe this will give you some ideas, though I don't know if this strategy is possible with Mithril's architecture :\

http://codepen.io/anon/pen/PZQpzN

[1] https://github.com/leeoniya/domvm

**Comment by vicnicius** [comment]
@isiahmeadows I agree with you. Sure this behaviour is an indicator of something wrong with the diffing algorithm or that it's just not as fault proof as it could be. Comparison with other libs reinforces that. 

What is not clear to me is if this was intended or not. As you said, handling such case would lead to some performance matters regarding memory use and the overall lib size. From my understanding (and notice that I'm just a lib user), mithril took a simpler approach to reconciliation (and to the whole api in general), and it was this simplicity that allowed it's small code base and it's low memory usage. Of course making it simpler would have it's pay-offs, like a fragile key diffing algorithm or the lack of hooks to component's rendering cycle. 

So, if the necessity of a stricter use of keys would allow relevant performance benefits, then this would not be a bug. If mithril wants to be more robust and handle such cases with relevant performance penalties, then it's a bug and mithril is changing it's direction. Finally, if there's a way to handle this case without performance penalties, then this is just a bug. 

Does that make any sense?

**Comment by vicnicius** [comment]
@leeoniya I took a quick look at domvm and really liked it! I'm willing to take some time to play with it on the next couple of weeks so that I can provide you some good feedback : )

**Comment by lhorie** [comment]
Mixing of keyed nodes and non-keyed nodes is properly supported in rewrite branch
